pkg/kubelet/server/server.go

```diff
@@ -258,6 +261,32 @@ func ListenAndServePodResources(ctx context.Context, endpoint string, providers
 	}
 }
 
+// ListenAndServePodsServer initializes an HTTP server to serve the Pod API.
+func ListenAndServePodsServer(ctx context.Context, endpoint string, srv podsv1alpha1.PodsServer) {
+	logger := klog.FromContext(ctx)
+	server := grpc.NewServer(apisgrpc.WithRateLimiter(ctx, "pods", pods.DefaultQPS, pods.DefaultBurstTokens))
+
+	podsv1alpha1.RegisterPodsServer(server, srv)
+
+	l, err := util.CreateListener(endpoint)
+	if err != nil {
+		logger.Error(err, "Failed to create listener for pods API endpoint")
+		os.Exit(1)
+	}
+
+	logger.Info("Starting to serve the pods API", "endpoint", endpoint)
+	go func() {
+		if err := server.Serve(l); err != nil && !errors.Is(err, grpc.ErrServerStopped) {
+			logger.Error(err, "Failed to serve")
+			os.Exit(1)
```

Comment:
do we exit kubelet here completely? Sounds harsh.

For device plugin we are doing health monitoring. This bug fix may give a pointer: https://github.com/kubernetes/kubernetes/pull/135153

I also think we delete-recreate a folder on kubelet restart. This may be useful here as well. We had good and bad from this behavior (removing the whole device plugin directory on restart). But the good thing is that some issues are remediated by the recreation of that directory.

cc @ffromani  again

Reply:
Fair enough.  This logic was shamelessly copied from PodResources as well, so we can patch both with whatever fix we decide on.